### PR TITLE
Wip 16998

### PR DESCRIPTION
**Pull request by athanatos**

### CODE REVIEW

File: src/osd/PG.cc [hunk 1]
```diff
@@ -2546,11 +2546,11 @@ void PG::_update_calc_stats()
 
     // a misplaced object is not stored on the correct OSD
     uint64_t misplaced = 0;
-    unsigned i = 0;
     unsigned in_place = 0;
-    for (vector<int>::iterator p = up.begin(); p != up.end(); ++p, ++i) {
-      pg_shard_t s(*p,
-		   pool.info.ec_pool() ? shard_id_t(i) : shard_id_t::NO_SHARD);
+    for (set<pg_shard_t>::const_iterator p = upset.begin();
```
Comment (tchaikov) [inline src/osd/PG.cc]: what is the difference between this approach and the existing way? i think they are the same.

Comment (tchaikov) [inline src/osd/PG.cc]: can we skip the shards who are `CRUSH_ITEM_NONE`? and could use `auto` and range based loop.

**Comment by tchaikov** [comment]
nit, might want to connect to http://tracker.ceph.com/issues/16998 using "Fixes: http://tracker.ceph.com/issues/16998"?

File: src/osd/PG.cc (hunk 1 above)
Comment (athanatos) [inline src/osd/PG.cc]: It's really meant to match the actingset and actingbackfill sets.  Mostly, using the up and acting vectors directly has led to a ton of subtle bugs like this one.

**Comment by athanatos** [comment]
Added the Fixes: line (thanks!).

File: src/osd/PG.cc (hunk 1 above)
Comment (tchaikov) [inline src/osd/PG.cc]: also, could use:
```
for (const auto &s : upset) {
```
here, and replace `*p` with `s` at https://github.com/ceph/ceph/pull/10688/files#diff-dfb9ddca0a3ee32b266623e8fa489626R2562 and  https://github.com/ceph/ceph/pull/10688/files#diff-dfb9ddca0a3ee32b266623e8fa489626R2567.

**Comment by tchaikov** [comment]
modulo a nit. lgtm.

**Comment by tchaikov** [comment]
retest this please.

File: src/osd/PG.cc (hunk 1 above)
Comment (athanatos) [inline src/osd/PG.cc]: Eh, I wanted it to be symmetric with the actingset scan above.  Also, we might want to backport these to hammer.

**Comment by tchaikov** [comment]
lgtm.

File: src/osd/PG.cc (hunk 1 above)
Comment (tchaikov) [inline src/osd/PG.cc]: okay.

**Comment by athanatos** [comment]
retest this please.

**Comment by yuriw** [comment]
http://pulpito.front.sepia.ceph.com:80/yuriw-2016-08-18_19:38:45-rados-wip-yuri-testing_2016_08_18-distro-basic-smithi/

**Pull request merged**